README.md
# React Placeholder
A React component to easily replicate your page with nice placeholders while the content is loading.
You can use a placeholder from the default set, or pass your own!

![image](https://cloud.githubusercontent.com/assets/691940/24140211/78406120-0e1f-11e7-9738-af2b2434c50e.png)

```jsx
import ReactPlaceholder from 'react-placeholder';
import "react-placeholder/lib/reactPlaceholder.css";

React.renderComponent(
  <div>
    <ReactPlaceholder type='media' rows={7} ready={this.state.ready}>
      <MyComponent />
    </ReactPlaceholder>
  </div>,
  document.body);
```

[**Live Demo**](http://buildo.github.io/react-placeholder/#!/ReactPlaceholder)

### Install
```
npm install --save react-placeholder
```

### Props

```tsx
children:              ReactNode;
ready:                 boolean;
delay?:                number;
firstLaunchOnly?:      boolean;
showLoadingAnimation?: boolean;
type?:                 'text' | 'media' | 'textRow' | 'rect' | 'round';
rows?:                 number;
color?:                string;
customPlaceholder?:    ReactElement;
className?:            string;
style?:                CSSProperties;
```

The default props will render a `text` placeholder with `3` rows and the color `#CDCDCD`.

### Customization
If the built-in set of placeholders is not enough, you can pass you own through the prop **"customPlaceholder"**

```jsx
<ReactPlaceholder ready={this.state.ready} customPlaceholder={<MyCustomPlaceholder />}>
  <MyComponent />
</ReactPlaceholder>
```

You can also import the built-in placeholders directly. This might be useful to use them to create your own customized placeholder:

```jsx
import {TextBlock, MediaBlock, TextRow, RectShape, RoundShape} from 'react-placeholder/lib/placeholders';

const awesomePlaceholder = (
  <div className='my-awesome-placeholder'>
    <RectShape color='blue' style={{width: 30, height: 80}}/>
    <TextBlock rows={7} color='yellow'/>
  </div>
);

<ReactPlaceholder ready={this.state.ready} customPlaceholder={awesomePlaceholder}>
  <MyComponent />
</ReactPlaceholder>
```

### Delay
You can pass an optional `delay` prop which specifies the time (in milliseconds) `react-placeholder` should wait before displaying the placeholder element. This is useful if you want to show a placeholder for slower connections while avoiding a brief "flash" on faster connections.

Note that this delay will __not__ affect the initial render, only subsequent "ready" state changes. Setting the `firstLaunchOnly` prop to `true` will also disable this feature.

### Animation
`react-placeholder` already comes with one default pulse animation to better tell the user that the page is loading.
The animation is defined in a separate CSS file so, in order to enable it, you should import that style in your project like this:

```js
import 'react-placeholder/lib/reactPlaceholder.css';
```

Once you've done this, you can simply pass the boolean prop `showLoadingAnimation` to tell `ReactPlaceholder` to animate itself:

```jsx
import 'react-placeholder/lib/reactPlaceholder.css';

<ReactPlaceholder showLoadingAnimation ready={this.state.ready} type="media" rows={5}>
  <p>This is a Test.</p>
</ReactPlaceholder>
```

### Style
you can style the placeholder by passing **```className```** or **```style```** or by using the built-in classes:

*"text-block", "media-block", "text-row", "rect-shape", "round-shape".*


CHANGELOG.md
#  Change Log



## [v4.1.0](https://github.com/buildo/react-placeholder/tree/v4.1.0) (2021-02-05)
[Full Changelog](https://github.com/buildo/react-placeholder/compare/v4.0.3...v4.1.0)

#### New features:

- Problem with React 17 and npm version  7.5.1 on mac osx [#95](https://github.com/buildo/react-placeholder/issues/95)
- [Question] CSS Import [#94](https://github.com/buildo/react-placeholder/issues/94)
- type CustomPlaceholderProps showLoadingAnimation is undefined instead of boolean [#90](https://github.com/buildo/react-placeholder/issues/90)
- v3.0.2 TypeScript error about missing properties [#82](https://github.com/buildo/react-placeholder/issues/82)
- Do `children` and `ready` really have to be required props? [#79](https://github.com/buildo/react-placeholder/issues/79)

## [v4.0.3](https://github.com/buildo/react-placeholder/tree/v4.0.3) (2020-06-11)
[Full Changelog](https://github.com/buildo/react-placeholder/compare/v4.0.2...v4.0.3)

## [v4.0.2](https://github.com/buildo/react-placeholder/tree/v4.0.2) (2020-06-01)
[Full Changelog](https://github.com/buildo/react-placeholder/compare/v4.0.1...v4.0.2)

## [v4.0.1](https://github.com/buildo/react-placeholder/tree/v4.0.1) (2020-03-13)
[Full Changelog](https://github.com/buildo/react-placeholder/compare/v4.0.0...v4.0.1)

#### New features:

- TextRow/TextBlock ignores lineSpacing is equal to 0 [#86](https://github.com/buildo/react-placeholder/issues/86)
- Is this package still maintained?  [#84](https://github.com/buildo/react-placeholder/issues/84)

## [v4.0.0](https://github.com/buildo/react-placeholder/tree/v4.0.0) (2020-02-29)
[Full Changelog](https://github.com/buildo/react-placeholder/compare/v3.0.2...v4.0.0)

#### Breaking:

- Warning: componentWillReceiveProps has been renamed, and is not recommended for use. [#83](https://github.com/buildo/react-placeholder/issues/83)

## [v3.0.2](https://github.com/buildo/react-placeholder/tree/v3.0.2) (2019-02-21)
[Full Changelog](https://github.com/buildo/react-placeholder/compare/v3.0.1...v3.0.2)

#### New features:

- ReactPlaceholder text component doesn't override the width property [#78](https://github.com/buildo/react-placeholder/issues/78)
- [Typings] Can't assign Element to customPlaceholder [#75](https://github.com/buildo/react-placeholder/issues/75)

## [v3.0.1](https://github.com/buildo/react-placeholder/tree/v3.0.1) (2018-05-08)
[Full Changelog](https://github.com/buildo/react-placeholder/compare/v3.0.0...v3.0.1)

#### Fixes (bugs & defects):

- Removed CSS? [#72](https://github.com/buildo/react-placeholder/issues/72)

## [v3.0.0](https://github.com/buildo/react-placeholder/tree/v3.0.0) (2018-05-04)
[Full Changelog](https://github.com/buildo/react-placeholder/compare/v2.0.0...v3.0.0)

#### Breaking:

- Refactor in TypeScript [#69](https://github.com/buildo/react-placeholder/issues/69)

#### New features:

- index.d.ts missing 'style' property [#67](https://github.com/buildo/react-placeholder/issues/67)

## [v2.0.0](https://github.com/buildo/react-placeholder/tree/v2.0.0) (2017-12-20)
[Full Changelog](https://github.com/buildo/react-placeholder/compare/v1.0.10...v2.0.0)

#### Breaking:

- Array.apply not support ie8 in TextBlock even if I add the polyfill (again) [#65](https://github.com/buildo/react-placeholder/issues/65)

## [v1.0.10](https://github.com/buildo/react-placeholder/tree/v1.0.10) (2017-12-15)
[Full Changelog](https://github.com/buildo/react-placeholder/compare/v1.0.9...v1.0.10)

#### Fixes (bugs & defects):

- Array.apply not support ie8 in TextBlock even if I'm add the polyfill [#62](https://github.com/buildo/react-placeholder/issues/62)

#### New features:

- TextBlock: make `widths` overridable [#59](https://github.com/buildo/react-placeholder/issues/59)

## [v1.0.9](https://github.com/buildo/react-placeholder/tree/v1.0.9) (2017-11-28)
[Full Changelog](https://github.com/buildo/react-placeholder/compare/v1.0.8...v1.0.9)

#### New features:

- Bug with delay when the component is already mounted [#54](https://github.com/buildo/react-placeholder/issues/54)

## [v1.0.8](https://github.com/buildo/react-placeholder/tree/v1.0.8) (2017-10-25)
[Full Changelog](https://github.com/buildo/react-placeholder/compare/v1.0.7...v1.0.8)

#### New features:

- Add "delay" prop to show the placeholder for a minimum time [#49](https://github.com/buildo/react-placeholder/issues/49)
- React 16 warning message [#46](https://github.com/buildo/react-placeholder/issues/46)

## [v1.0.7](https://github.com/buildo/react-placeholder/tree/v1.0.7) (2017-10-16)
[Full Changelog](https://github.com/buildo/react-placeholder/compare/v1.0.6...v1.0.7)

#### New features:

- error in ready [#44](https://github.com/buildo/react-placeholder/issues/44)

## [v1.0.6](https://github.com/buildo/react-placeholder/tree/v1.0.6) (2017-08-03)
[Full Changelog](https://github.com/buildo/react-placeholder/compare/v1.0.5...v1.0.6)

#### Fixes (bugs & defects):

- @types/react > 33 complains [#41](https://github.com/buildo/react-placeholder/issues/41)
- Animations not working with custom placeholders. [#34](https://github.com/buildo/react-placeholder/issues/34)

#### New features:

- Action required: Greenkeeper could not be activated 🚨 [#39](https://github.com/buildo/react-placeholder/issues/39)
- Action required: Greenkeeper could not be activated 🚨 [#38](https://github.com/buildo/react-placeholder/issues/38)

## [v1.0.5](https://github.com/buildo/react-placeholder/tree/v1.0.5) (2017-06-26)
[Full Changelog](https://github.com/buildo/react-placeholder/compare/v1.0.4...v1.0.5)

#### Fixes (bugs & defects):

- Typo in CSS selector for react-shape animation [#33](https://github.com/buildo/react-placeholder/issues/33)

## [v1.0.4](https://github.com/buildo/react-placeholder/tree/v1.0.4) (2017-06-01)
[Full Changelog](https://github.com/buildo/react-placeholder/compare/v1.0.3...v1.0.4)

#### Fixes (bugs & defects):


examples/examples.js
import React from 'react';
import ReactDOM from 'react-dom';
import ReactPlaceholder from '../src';
import {
  TextBlock,
  MediaBlock,
  TextRow,
  RectShape,
  RoundShape
} from '../src/placeholders';

import '../src/reactPlaceholder.scss';

class Example extends React.Component {

  state = {
    delay: 1000,
    ready: false,
    readyCustom: false,
    readyFirstLaunch: false,
    textBlockRows: 6,
    mediaBlockRows: 4,
    holdPlaceholder: 2000
  };

  onChange = (key) => ({ target: { value } }) => {
    this.setState({ [key]: parseInt(value) || '' });
  }

  toggleReady = () => {
    this.setState({ ready: !this.state.ready });
  }

  toggleReadyCustom = () => {
    this.setState({ readyCustom: !this.state.readyCustom });
  }

  toggleReadyFirstLaunch = () => {
    this.setState({ readyFirstLaunch: !this.state.readyFirstLaunch });
  }

  render() {
    const buttonStyle = {
      marginBottom: 10
    };

    const customPlaceholder = (
      <div style={{ color: 'blue', backgroundColor: 'yellow' }}>
        I'm a custom filler!!
      </div>
    );

    const realComponent = (
      <div style={{ padding: 10, backgroundColor: '#F0F0F0', color: 'red' }}>
        I'm the real component!
      </div>
    );

    return (
      <div style={{ margin: 30 }}>
        <h1>Using ReactPlaceholder types</h1>
        <button onClick={this.toggleReady} style={buttonStyle}>
          {this.state.ready ? 'set loading' : 'set ready'}
        </button>
        <p>"text"</p>
        <div className='ui input'>
          <span style={{ lineHeight: '40px' }}>
            nº of rows:
          </span>
          <input
            type='number'
            value={this.state.textBlockRows || ''}
            onChange={this.onChange('textBlockRows')}
            style={{ width: 80, marginLeft: 5 }}
          />
        </div>
        <div className='ui segment'>
          <ReactPlaceholder
            ready={this.state.ready}
            rows={this.state.textBlockRows}
            color='#E0E0E0'
            className='my-text-block'
          >
            {realComponent}
          </ReactPlaceholder>
        </div>

        <p>"media (with loading animation)"</p>
        <div className='ui input'>
          <span style={{ lineHeight: '40px' }}>
            nº of rows:
          </span>
          <input
            type='number'
            value={this.state.mediaBlockRows || ''}
            onChange={this.onChange('mediaBlockRows')}
            style={{ width: 80, marginLeft: 5 }}
          />
        </div>
        <div className='ui segment'>
          <ReactPlaceholder
            showLoadingAnimation
            ready={this.state.ready}
            type='media'
            rows={this.state.mediaBlockRows}
            className='my-text-block'
          >
            {realComponent}
          </ReactPlaceholder>
        </div>

        <p>"textRow"</p>
        <div className='ui segment'>
          <ReactPlaceholder
            ready={this.state.ready}
            type='textRow'
            color='#E0E0E0'
            className='my-text-row'
          >
            {realComponent}
          </ReactPlaceholder>
        </div>

        <p>"rect"</p>
        <div className='ui segment'>
          <ReactPlaceholder
            ready={this.state.ready}
            type='rect'
            color='#E0E0E0'
            style={{ width: 50, height: 50 }}
            className='my-rect'
          >
            {realComponent}
          </ReactPlaceholder>
        </div>

        <p>"round"</p>
        <div className='ui segment'>
          <ReactPlaceholder
            ready={this.state.ready}
            type='round'
            color='#E0E0E0'
            style={{ width: 50, height: 50 }}
            className='my-round'
          >
            {realComponent}
          </ReactPlaceholder>
        </div>

        <h1>passing a customPlaceholder</h1>
        <button onClick={this.toggleReadyCustom} style={buttonStyle}>
          {this.state.readyCustom ? 'set loading' : 'set ready'}
        </button>
        <div className='ui segment'>
          <ReactPlaceholder
            ready={this.state.readyCustom}
            customPlaceholder={customPlaceholder}
          >
            {realComponent}
          </ReactPlaceholder>
        </div>

        <h1>Using ReactPlaceholder with firstLaunchOnly=true</h1>
        <button onClick={this.toggleReadyFirstLaunch} style={buttonStyle}>
          {this.state.readyFirstLaunch ? 'set loading' : 'set ready'}
        </button>
        <p>Will show filler only once</p>
        <div className='ui segment'>
          <ReactPlaceholder
            ready={this.state.readyFirstLaunch}
            rows={4}
            color='#E0E0E0'
            className='my-text-block'
            firstLaunchOnly
          >
            {realComponent}
          </ReactPlaceholder>
        </div>

        <h1>Using ReactPlaceholder with a hold placeholder time</h1>
        <button onClick={this.toggleReadyFirstLaunch} style={buttonStyle}>
          {this.state.readyFirstLaunch ? 'set loading' : 'set ready'}
        </button>
        <p>Will show the content only after specified time and it's ready</p>
        <div className='ui input'>
          <span style={{ lineHeight: '40px' }}>
            holdPlaceholder (ms):
          </span>
          <input
            type='number'
            value={this.state.holdPlaceholder}
            onChange={this.onChange('holdPlaceholder')}
            style={{ width: 80, marginLeft: 5 }}
          />
        </div>
        <div className='ui segment'>
          <ReactPlaceholder
            ready={this.state.readyFirstLaunch}
            holdPlaceholder={this.state.holdPlaceholder}
            rows={4}
            color='#E0E0E0'
            className='my-text-block'
          >
            {realComponent}
          </ReactPlaceholder>
        </div>

        <h1>Using ReactPlaceholder with a delay</h1>
        <button onClick={this.toggleReadyFirstLaunch} style={buttonStyle}>
          {this.state.readyFirstLaunch ? 'set loading' : 'set ready'}
        </button>
        <p>Will show filler only after specified delay</p>
        <div className='ui input'>
          <span style={{ lineHeight: '40px' }}>
            delay (ms):
          </span>
          <input
            type='number'
            value={this.state.delay}
            onChange={this.onChange('delay')}
            style={{ width: 80, marginLeft: 5 }}
          />
        </div>
        <div className='ui segment'>
          <ReactPlaceholder
            ready={this.state.readyFirstLaunch}
            delay={this.state.delay}
            rows={4}
            color='#E0E0E0'
            className='my-text-block'
          >
            {realComponent}
          </ReactPlaceholder>
        </div>

        <h1>Using default placeholders directly</h1>
        <p>They don't accept children</p>
        <p>TextBlock</p>
        <div className='ui segment'>
          <TextBlock
            color='#E0E0E0'
            rows={4}
            className='my-text-row'
          />
        </div>
        <p>MediaBlock</p>
        <div className='ui segment'>
          <MediaBlock
            color='#E0E0E0'
            rows={4}
            className='my-text-row'
          />
        </div>
        <p>TextRow</p>
        <div className='ui segment'>
          <TextRow
            color='#E0E0E0'
            className='my-text-row'
          />
        </div>

        <p>RectShape</p>
        <div className='ui segment'>
          <RectShape
            color='#E0E0E0'
            className='my-text-row'
            style={{ width: 50, height: 50 }}
          />
        </div>

        <p>RoundShape</p>
        <div className='ui segment'>
          <RoundShape
            color='#E0E0E0'
            className='my-text-row'
            style={{ width: 50, height: 50 }}
          />
        </div>

      </div>
    );
  }

}

ReactDOM.render(<Example />, document.getElementById('container'));
